### PR TITLE
Add a record layer design document

### PR DESCRIPTION
**Pull request by mattcaswell**

### CODE REVIEW

**Comment by mattcaswell** [comment]
Added the "need OTC decision" label since this needs discussion at OTC.

File: doc/designs/quic-design/record-layer.md [hunk 1]
```diff
@@ -0,0 +1,470 @@
+Design Problem: Abstract Record Layer
+=====================================
+
+This document covers the design of an abstract record layer for use in (D)TLS
+and QUIC.
+
+A record within this document refers to a packet of data. It will typically
+contain some form of header data and some payload data, and will often be
+cryptographically protected. A record may or may have a one-to-one
```
Comment (slontis) [inline doc/designs/quic-design/record-layer.md:9]: may not ?

**Comment by paulidale** [comment]
Is this ready for review?

**Comment by levitte** [comment]
Is this still in progress?  Considering that #18132, #18610 and #18795 have been merged, does that indicate that this has fulfilled its purpose?  Should it be merged or just closed?

**Comment by mattcaswell** [comment]
This is the basis I am using for the implementation. There are already numerous changes from what is described here implemented in #18132. My intention is to circle back and update this design document with the latest changes once things have settled down with the implementation of the write side.

**Comment by paulidale** [comment]
I'd much prefer to have seen this merged and updated separately.
Anything that depends on the record layer is currently in limbo.

**Comment by mattcaswell** [comment]
This PR has now been updated to reflect the current state of the record layer as it has been implemented.

I have taken it out of draft and it is now ready for review.

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
@mattcaswell Some MD nits.

**Comment by mattcaswell** [comment]
Fixup pushed to address md-nits issues. @hlandau, @t8m please reconfirm.

**Review by t8m — APPROVED** [review]

**Review by hlandau — APPROVED** [review]
Still OK

**Comment by hlandau** [comment]
CIs are not relevant.

Merged to master. Thank you.

**Pull request closed without merging**